### ISSUE TITLE
find out why learning stops sometimes

### ISSUE THREAD

**Issue by cwmeijer** [NLeSC/mcfly#43]
Sometimes learning stops with a completely stable accuracy and loss on val and test set. Find out why the process is stuck.
- [x] Find cases when this is happening and save them so they can be replicated
- [ ] Find out why it is stuck in the current state (0 gradient?)
- [ ] Find out how the process got to the current state (vanishing or exploding gradient? inappropriate learning rate?)

**Comment by dafnevk** [comment]
Here are the results of 10 models on the PAMAP2 dataset:
![image](https://cloud.githubusercontent.com/assets/4925047/17289529/68a30d7c-57da-11e6-89ff-c5acc9144c4b.png)
In the training process plots, it's clearly visible that the last 2 models don't learn.
Theses two models have a learning rate of >0.03 and none of the other models have such a high learning rate. So my suspicion is that the high learning rate has to do with it.

**Comment by dafnevk** [comment]
I think the loss=nan problem is something different. The learning than really stops, because we get NaN weights (and thus NaN loss).
I ran into this problem with the Pamap2 dataset (when training on the complete train set).
I'm trying to find out why. See for example https://groups.google.com/forum/#!topic/theano-users/UTn3hepy1sw.

**Comment by dafnevk** [comment]
This is a useful resource: http://russellsstewart.com/notes/0.html

I tried overfitting a network on a very small subset of the EEG dataset, This works for most models but if the learning rate is very small then it just takes very many iterations before it overfits.
If the learning rate is too large, then it won't overfit. and it just gets stuck at a certain loss.

**Comment by dafnevk** [comment]
For the NaN loss in the Pamap2 dataset, I think it has to do with an overflow in computing the categorical cross-entropy loss.
According to [this issue](https://github.com/fchollet/keras/issues/1429#issuecomment-181426156) this can be solved by adjusting some settings in Theano.
I tried setting K.epsilon to 1e-5 (which already seems quite large to me) and this just prolonged it to the 5th epoch and then the loss became NaN again.

The model that gave this problem (the best model among the 10 that I tried) was a CNN  with hyperparams:
{'filters': array([98, 67]), 'fc_hidden_nodes': 62, 'regularization_rate': 0.03708604649827486, 'learning_rate': 0.00019850152863738226}

**Comment by dafnevk** [comment]
The following CNN model produces NaN in the first epoch:
Numpy Seed: 123
All Pamap2 train data
params = {'filters': np.array([98, 67]), 'fc_hidden_nodes': 62, 'regularization_rate': 0.03708604649827486, 'learning_rate': 0.00019850152863738226}

**Comment by cwmeijer** [comment]
It looks like on my computer I get NaN for loss for different seeds. I wrote a loop to find such cases in the PAMAP debugging notebook. I found cases where I get a NaN loss with only 1 epoch and 100 data points.
I was also able to set the NanGuardMode for Theano to throw exceptions for NaN, and not for Inf or big values. I couldn't manage this by just configuring Theano, even without using Keras. If I set the theano.config.mode = 'NanGuardMode', it will stop at any of these three cases, even if I set theano.config.NanGuardMode.inf_is_error = False.
The trick I had to use was to add one line to the Keras code. In 
keras/backend/theano_backend.py:514 change the function to:
def __init__(self, inputs, outputs, updates=[], **kwargs):
        self.function = theano.function(inputs, outputs, updates=updates,
                                        allow_input_downcast=True,
                                        on_unused_input='warn',
                                        mode=theano.compile.nanguardmode.NanGuardMode(),
                                        **kwargs)
This worked for me. You could also add arguments but I didn't need to so I didn't try that:
NanGuardMode(inf_is_error=True, big_is_error=None, nan_is_error=None)

**Comment by cwmeijer** [comment]
It was suggested that too low of a regularization rate could cause the NaNs. I trained a model with the same seed that caused a NaN val_loss, with the same dataset of 1 training example and 1 validation example. The result was that all values had changed (higher losses when using ridiculously high reg rates) except for the val_loss which was NaN before, and still is NaN with the higher regularization.
